### ISSUE TITLE
kubelet restarts container after it finished successfully even with restartPolicy Never

### ISSUE THREAD

**Issue by cezarsa** [kubernetes/kubernetes#51142]
/kind bug

**What happened**:

After creating a Pod with `restartPolicy: Never` and 2 containers, if one of the containers has finished successfully and was removed (either by `docker rm` or by the GC), kubelet will restart the container.

**What you expected to happen**:

Since the container already executed and finished successfully, and the restart policy was set to Never, it was expected that the container could be safely removed without side-effects.

**How to reproduce it (as minimally and precisely as possible)**:

1. Create a Pod with:
```
apiVersion: v1
kind: Pod
metadata:
  name: mypod
  namespace: default
spec:
  containers:
  - command: ["sh", "-c", "sleep 1000"]
    image: alpine
    name: sleeper
  - command: ["sh", "-c", "exit 0"]
    image: alpine
    name: completed
  restartPolicy: Never
```
2. Connect to the node where the Pod was scheduled. This is the expected state:
```
# docker ps -a | grep mypod
588f18895f80 alpine@sha256:... "sh -c 'sleep 1000'" 14 seconds ago Up 13 seconds                                   k8s_sleeper_mypod_default_42a2a5ae-876c-11e7-921e-065bde0090ff_0
b17915ea38df alpine@sha256:... "sh -c 'exit 0'" 14 seconds ago Exited (0) 13 seconds ago                       k8s_completed_mypod_default_42a2a5ae-876c-11e7-921e-065bde0090ff_0
```
3. Run `docker rm b17915ea38df` (ID of the container with "Exited (0)" status)
4. Wait about a minute and check the containers again
```
# docker ps -a | grep mypod
469b49300eb3 alpine@sha256:... "sh -c 'exit 0'" 4 seconds ago Exited (0) 3 seconds ago                       k8s_completed_mypod_default_42a2a5ae-876c-11e7-921e-065bde0090ff_0
588f18895f80 alpine@sha256:... "sh -c 'sleep 1000'" About a minute ago Up About a minute                              k8s_sleeper_mypod_default_42a2a5ae-876c-11e7-921e-065bde0090ff_0
```
5. We can see that the container with the "exit 0" command was created again with a new Docker ID. Also, it's possible to see the following message in `kubelet.log` file:
> Container {Name:completed Image:alpine Command:[sh -c exit 0] Args:[] WorkingDir: Ports:[] EnvFrom:[] Env:[] Resources:{Limits:map[] Requests:map[]} VolumeMounts:[{Name:default-token-1l316 ReadOnly:true MountPath:/var/run/secrets/kubernetes.io/serviceaccount SubPath:}] LivenessProbe:nil ReadinessProbe:nil Lifecycle:nil TerminationMessagePath:/dev/termination-log TerminationMessagePolicy:File ImagePullPolicy:Always SecurityContext:nil Stdin:false StdinOnce:false TTY:false} is dead, **but RestartPolicy says that we should restart it**.

This message even states that `RestartPolicy says that we should restart it` which is obviously false. 

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): 1.7.4
```
Client Version: version.Info{Major:"1", Minor:"7", GitVersion:"v1.7.4", GitCommit:"793658f2d7ca7f064d2bdf606519f9fe1229c381", GitTreeState:"clean", BuildDate:"2017-08-17T08:48:23Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"7", GitVersion:"v1.7.4", GitCommit:"793658f2d7ca7f064d2bdf606519f9fe1229c381", GitTreeState:"clean", BuildDate:"2017-08-17T08:30:51Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration**: Cloudstack private cloud.
- OS (e.g. from /etc/os-release): `Ubuntu 14.04.5 LTS`
- Kernel (e.g. `uname -a`): `Linux kubernetes3 3.13.0-107-generic #154-Ubuntu SMP Tue Dec 20 09:57:27 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux`

**Comment by xingzhou** [comment]
seems only related to multi-container pod? I'd like to try and investigate this

**Comment by xingzhou** [comment]
can be reproduced on latest Master code, in my env, after about 1 minute, the removed container will be run again. Looking into the code right now.

**Comment by xingzhou** [comment]
these lines show the reason why kubelet keeps restart the finished container: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/container/helpers.go#L66-L71

The reason here is when a container exits and removed in docker, docker can not get the status of this container any more, so the status object for this container becomes nil, which means the container does not ever started from a kubelet's view.

So if the container's last State is `Exited` and the container is `removed` by GC or user, we need to keep the last non-nil container's status so that in next `syncPod` of kubelet, this container can be restarted according to the pod's restart policy.

These lines also indicate that kubelet can detect the `removed` container:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/pleg/generic.go#L141-L165

I'm submitting a patch to fix part of this issue, for discussion

**Comment by cezarsa** [comment]
I was still able to reproduce the issue following the steps in the description using Kubernetes v1.9.1.
I can see that @xingzhou's fix in #51309 was closed automatically and had some failing tests. If I find some time later this month I may be able to help rebasing it and checking the tests.

/remove-lifecycle stale
